Kernel/Syscall.cpp

```diff
@@ -138,88 +94,4 @@ ErrorOr<FlatPtr> handle(RegisterState& regs, FlatPtr function, FlatPtr arg1, Fla
 
 }
 
-NEVER_INLINE void syscall_handler(TrapFrame* trap)
-{
-    // Make sure SMAP protection is enabled on syscall entry.
-    clac();
-
-    auto& regs = *trap->regs;
-    auto* current_thread = Thread::current();
-    VERIFY(current_thread->previous_mode() == Thread::PreviousMode::UserMode);
-    auto& process = current_thread->process();
-    if (process.is_dying()) {
-        // It's possible this thread is just about to make a syscall while another is
-        // is killing our process.
-        current_thread->die_if_needed();
-        return;
-    }
-
-    if (auto* tracer = process.tracer(); tracer && tracer->is_tracing_syscalls()) {
-        tracer->set_trace_syscalls(false);
-        process.tracer_trap(*current_thread, regs); // this triggers SIGTRAP and stops the thread!
-    }
-
-    current_thread->yield_if_stopped();
-
-    // Apply a random offset in the range 0-255 to the stack pointer,
-    // to make kernel stacks a bit less deterministic.
-    u32 lsw;
-    u32 msw;
-    read_tsc(lsw, msw);
-
-    auto* ptr = (char*)__builtin_alloca(lsw & 0xff);
-    asm volatile(""
-                 : "=m"(*ptr));
-
-    constexpr FlatPtr iopl_mask = 3u << 12;
-
-    FlatPtr flags = regs.flags();
-    if ((flags & (iopl_mask)) != 0) {
-        PANIC("Syscall from process with IOPL != 0");
-    }
-
-    Memory::MemoryManager::validate_syscall_preconditions(process, regs);
-
-    FlatPtr function;
-    FlatPtr arg1;
-    FlatPtr arg2;
-    FlatPtr arg3;
-    FlatPtr arg4;
-    regs.capture_syscall_params(function, arg1, arg2, arg3, arg4);
-
-    auto result = Syscall::handle(regs, function, arg1, arg2, arg3, arg4);
-
-    if (result.is_error()) {
-        regs.set_return_reg(-result.error().code());
-    } else {
-        regs.set_return_reg(result.value());
-    }
-
-    if (auto* tracer = process.tracer(); tracer && tracer->is_tracing_syscalls()) {
-        tracer->set_trace_syscalls(false);
-        process.tracer_trap(*current_thread, regs); // this triggers SIGTRAP and stops the thread!
-    }
-
-    current_thread->yield_if_stopped();
-
-    current_thread->check_dispatch_pending_signal();
-
-    // If the previous mode somehow changed something is seriously messed up...
-    VERIFY(current_thread->previous_mode() == Thread::PreviousMode::UserMode);
-
-    // Check if we're supposed to return to userspace or just die.
-    current_thread->die_if_needed();
-
-    // Crash any processes which have committed a promise violation during syscall handling.
-    if (result.is_error() && result.error().code() == EPROMISEVIOLATION) {
-        VERIFY(current_thread->is_promise_violation_pending());
-        current_thread->set_promise_violation_pending(false);
-        process.crash(SIGABRT, 0);
-    } else {
-        VERIFY(!current_thread->is_promise_violation_pending());
-    }
-
-    VERIFY(!g_scheduler_lock.is_locked_by_current_processor());
-}
```

Comment:
Most of this code works fine for aarch64, and should be shared between architectures. I have a patch locally that makes this file work for aarch64 by adding some #ifdefs. The only x86_64 specific things are the `clac()` and the code that adds the random offset to the stack and checks the iopl flags, the rest can and should be shared with aarch64.

Wasn't able to select the whole code block the my phone..

Reply:
Hello, thanks for the review. Makes sense, I will move this back to its place then and update the PR.

Reply:
Probably makes sense to wait for #17183 to go in first.

Reply:
Hello @FireFox317 and @nico I've rebased the PR and now #17183 is in. Also I've moved back the function that is platform agnostic.

Let me know if there is anything else I can do. Thanks.